### PR TITLE
charset: add a config field to control charset feature

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: https://github.com/pingcap/tidb/issues/26812

a config to control charset feature

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
add syntax support for gbk charset in variable, ddl, etc.
```

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -548,6 +548,8 @@ type IsolationRead struct {
 type Experimental struct {
 	// Whether enable global kill.
 	EnableGlobalKill bool `toml:"enable-global-kill" json:"-"`
+	// Whether enable charset feature.
+	EnableCharsetFeat bool `toml:"enable-charset-feat" json:"-"`
```
Comment (zimulala) [inline config/config.go:552]: Could we replace the configuration with a hidden global variable？ After this feature GA, we  can delete the variable directly.

Comment (xiongjiwei) [reply config/config.go:552]: I think we can not avoid it, this config should not modify online. I suggest using the config field but not expose to users, after the feature GA, we delete it. It is expected that this config is transparent for all.

Comment (zimulala) [reply config/config.go:552]: Well, I think that's ok.

**Review by tangenta — APPROVED** [review]

File: util/collate/charset.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+func CharsetFeatEnabled() bool {
+	return enableCharsetFeat
+}
+
+func addCharset() {
+	for _, c := range experimentalCharsetInfo {
+		charset.AddCharset(c)
+		for _, coll := range charset.GetCollations() {
+			if strings.EqualFold(coll.CharsetName, c.Name) {
+				charset.AddCollation(coll)
```
Comment (zimulala) [inline util/collate/charset.go:55]: Could we remove this code？
`GetCollations` gets collations from `charset.collations` and `AddCollation` adds a collation to `charset.collations`.

File: util/collate/charset.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"strings"
+
+	"github.com/pingcap/parser/charset"
+)
+
+var (
+	enableCharsetFeat bool
+)
+
+// EnableCharsetFeat the charset feature
```
Comment (zimulala) [inline util/collate/charset.go:27]:
```suggestion
// EnableCharsetFeat enables the charset feature.
```

File: util/collate/charset.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+func EnableCharsetFeat() {
+	enableCharsetFeat = true
+	addCharset()
+}
+
+// SetCharsetFratEnabledForTest set charset feature enabled. Only used in test.
+// It will also enable or disable new collation.
+func SetCharsetFratEnabledForTest(flag bool) {
```
Comment (zimulala) [inline util/collate/charset.go:35]: s/SetCharsetFratEnabledForTest/SetCharsetFeatEnabledForTest ?

File: util/collate/charset.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+	for name := range experimentalCollation {
+		delete(newCollatorMap, name)
+		delete(newCollatorIDMap, CollationName2ID(name))
+	}
+}
+
+// All the experimental supported charset should be in the following table, only used when charset feature is enable.
+var experimentalCharsetInfo = []*charset.Charset{
+	{Name: charset.CharsetGBK, DefaultCollation: charset.CollationGBKBin, Collations: make(map[string]*charset.Collation), Desc: "Chinese Internal Code Specification", Maxlen: 4},
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply util/collate/charset.go]: right, updated

File: executor/executor_test.go [hunk 6]
```diff
@@ -5627,6 +5628,66 @@ func (s *testSerialSuite2) TestUnsignedFeedback(c *C) {
...
+func (s *testSerialSuite2) TestCharsetFeature(c *C) {
+	collate.SetCharsetFratEnabledForTest(true)
+	defer collate.SetCharsetFratEnabledForTest(false)
+
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustQuery("show charset").Check(testkit.Rows(
+		"ascii US ASCII ascii_bin 1",
+		"binary binary binary 1",
+		"gbk Chinese Internal Code Specification gbk_bin 4",
```
Comment (zimulala) [inline executor/executor_test.go:5640]:
```suggestion
		"gbk Chinese Internal Code Specification gbk_bin 2",
```

File: util/collate/charset.go (hunk 2 above)
Comment (xiongjiwei) [reply util/collate/charset.go:55]: nope, `AddCollation` not only add it to `charset.collations`, it also add collation to `Charset.Collations` field.

Comment (zimulala) [reply util/collate/charset.go:55]: Then, do we need to remove it when we do `removeCharset`?

File: executor/executor_test.go [hunk 7]
```diff
@@ -5627,6 +5628,66 @@ func (s *testSerialSuite2) TestUnsignedFeedback(c *C) {
...
+	))
+	tk.MustExec("create table t4(a char(10));")
+	tk.MustExec("alter table t4 add column b char(10) charset gbk;")
+	tk.MustQuery("show create table t4").Check(testkit.Rows("t4 CREATE TABLE `t4` (\n" +
+		"  `a` char(10) DEFAULT NULL,\n" +
+		"  `b` char(10) CHARACTER SET gbk COLLATE gbk_bin DEFAULT NULL\n" +
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin",
+	))
+
+	tk.MustExec("create database test_gbk charset gbk;")
```
Comment (zimulala) [inline executor/executor_test.go:5688]: Maybe we can add `use test_gbk; create table t(a char(10));` to check the result of `show create table t`.

File: executor/executor_test.go [hunk 8]
```diff
@@ -5627,6 +5628,66 @@ func (s *testSerialSuite2) TestUnsignedFeedback(c *C) {
...
+	tk.MustQuery("select @@character_set_connection;").Check(testkit.Rows("gbk"))
+	tk.MustQuery("select @@collation_connection;").Check(testkit.Rows("gbk_bin"))
+	tk.MustExec("set @@character_set_client=gbk;")
+	tk.MustQuery("select @@character_set_client;").Check(testkit.Rows("gbk"))
+	tk.MustExec("set names utf8mb4;")
+	tk.MustExec("set @@character_set_connection=gbk;")
+	tk.MustQuery("select @@character_set_connection;").Check(testkit.Rows("gbk"))
+	tk.MustQuery("select @@collation_connection;").Check(testkit.Rows("gbk_bin"))
+
+	tk.MustExec("select _gbk 'a'")
```
Comment (zimulala) [inline executor/executor_test.go:5668]: Please check the result.

File: util/collate/charset.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+	"github.com/pingcap/parser/charset"
+)
+
+var (
+	enableCharsetFeat bool
+)
+
+// EnableCharsetFeat enables the charset feature.
+func EnableCharsetFeat() {
```
Comment (zimulala) [inline util/collate/charset.go:28]: Do we need to change it to `EnableNewCharset`? This matches the collation feature.

File: config/config.go [hunk 10]
```diff
@@ -669,7 +671,8 @@ var defaultConf = Config{
 		Engines: []string{"tikv", "tiflash", "tidb"},
 	},
 	Experimental: Experimental{
-		EnableGlobalKill: false,
+		EnableGlobalKill:  false,
+		EnableCharsetFeat: false,
```
Comment (zimulala) [inline config/config.go:675]: Please add a test for this configuration item.

File: util/collate/charset.go (hunk 2 above)
Comment (zimulala) [reply util/collate/charset.go:55]: But we have done `AddCollation` in `init` for all `charset.collations`(include gbk).

File: config/config_test.go [hunk 11]
```diff
@@ -286,6 +286,7 @@ deadlock-history-collect-retryable = true
 	require.Equal(t, uint64(30), conf.StoresRefreshInterval)
 	require.Equal(t, uint(123), conf.PessimisticTxn.DeadlockHistoryCapacity)
 	require.True(t, conf.PessimisticTxn.DeadlockHistoryCollectRetryable)
+	require.False(t, conf.Experimental.EnableNewCharset)
```
Comment (xiongjiwei) [inline config/config_test.go:289]: if no this field in config file, it should be false

File: util/collate/charset.go (hunk 2 above)
Comment (xiongjiwei) [reply util/collate/charset.go:55]: if charset is removed, it is removed.

Comment (zimulala) [reply util/collate/charset.go:55]: It seems that `charset.supportedCollations` doesn't remove it, when we do `show collation`(after `removeCharset`), the result is not what we expected.

Comment (xiongjiwei) [reply util/collate/charset.go:55]: `supportedCollations` is not changed in `AddCollation`, when `AddCollation` only the collation in `supportedCollationNames` will be added into `supportedCollations`.

File: util/collate/charset.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+var (
+	enableCharsetFeat bool
+)
+
+// EnableNewCharset enables the charset feature.
+func EnableNewCharset() {
+	enableCharsetFeat = true
+	addCharset()
+}
```
Comment (tangenta) [inline util/collate/charset.go:31]: A small suggestion: how about making the `addCharset()` idempotent?

Comment (xiongjiwei) [reply util/collate/charset.go:31]: it is, because the data is added into `map`

File: util/collate/charset.go (hunk 2 above)
Comment (zimulala) [reply util/collate/charset.go:55]: Got it.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**